doc/design/context.md

```diff
@@ -0,0 +1,171 @@
+## Context Design
+
+`Net` is the container and controller of a set of `Operator`. Each `Operator` in `Net` has a method called `Run` to make computation. The `Run` method of `Operator` is defined as follows:
+
+```c++
+Error Operator::Run(OpContext* context);
+```
+
+And the `OpContext` is defined as follows:
+
+```c++
+Struct OpContext {
+  Scope* scope;
+  DeviceContext* device_context;
+};
+```
+
+The `Run` method will take `Variable` (containing `Tensor`) from `Scope` to make computation on certain `DeviceContext`. `DeviceContext` provides necessary runtime resources for computation, including CudaStream, cublasHandle and so on.
+
+Generally speaking, the Tensor computation are executed by Eigen library, which needs an Eigen::GpuDevice type object as parameter in a GPU card, and the GpuDevice parameter is constructed with an Eigen::CudaStreamDevice object. We need to set both a specific GpuID and CudaStream to create a Eigen::CudaStream object.
+
+At the same time, some computation work will executed by cuBLAS or cuDNN library. Take cuBLAS library as an example, we have to acquire a cublasHandle which binds on a CudaStream to make computation. It's the same way as Eigen library does.
+
+`DeviceContext` is defined as follows(just using for unify class `CudaDeviceContext` and class `CpuDeviceContext`):
+
+```
+class DeviceContext {
+  virtual ~Context() {}
+};
+```
+
+### CudaDeviceContext
+
+CudaDeviceContext is defined as follows:
+
+
+```c++                                   
+class DeviceGuard {
+ public:
+  explicit DeviceGuard(GPUPlace new_place)
+      : previous_(GetCurrentGPUID()) {
+    if (previous_ != new_place) {
+      cudaError_t err = cudaSetDevice(new_place.device);
+      PADDLE_ENFORCE(err == cudaSuccess);
+    }
+  }
+
+  ~DeviceGuard() {
+    cudaError_t err = cudaSetDevice(previous_.device);
+    PADDLE_ENFORCE(err == cudaSuccess);
+  }
+
+ private:
+  GPUPlace previous_;
+};
+
+class CudaDeviceContext : public DeviceContext{
+public:
+  explicit CDUAContext(const GPUPlace gpu_place) : gpu_place_(gpu_place) {
+    DeviceGuard(gpu_place_);
+    cudaError_t err = cudaStreamCreate(&stream_);
+    PADDLE_ENFORCE(err == cudaSuccess);
+    
+    eigen_stream_ = new Eigen::CudaStreamDevice(&stream_);
+    eigen_handle_ = new Eigen::GpuDevice(eigen_stream_);    
+  }
+  
+  void Wait() {
```

Comment:
This two-line function is only called by the destructor, how about move these two lines into the destructor and save this function?